eventsourcing/src/test/java/org/axonframework/eventsourcing/eventstore/DefaultEventStoreTransactionTest.java

```diff
@@ -0,0 +1,354 @@
+                   transaction.appendEvent(event1);
+                   transaction.appendEvent(event2);
+               })
+               .runOnPrepareCommit(context -> {
+                   throw new IllegalStateException("Simulated failure during prepare commit");
+               });
+
+            // then
+            assertThrows(CompletionException.class, () -> awaitException(uow.execute()));
+
+            var verificationUow = new AsyncUnitOfWork();
+            var eventsAfterRollback = new AtomicReference<MessageStream<? extends EventMessage<?>>>();
+            verificationUow.runOnPreInvocation(context -> {
+                EventStoreTransaction transaction = defaultEventStoreTransactionFor(context);
+                eventsAfterRollback.set(transaction.source(sourcingCondition, context));
+            });
+            awaitCompletion(verificationUow.execute());
+
+            StepVerifier.create(eventsAfterRollback.get().asFlux())
+                        .verifyComplete();
+        }
+
+        @Test
+        void errorPropagationIsHandledByOnErrorPhase() {
+            // given
+            var event1 = eventMessage(0);
+            var capturedError = new AtomicReference<Throwable>();
+            var onCommitExecuted = new AtomicBoolean(false);
+            var onAfterCommitExecuted = new AtomicBoolean(false);
+            var onPostInvocationExecuted = new AtomicBoolean(false);
+
+            // when
+            var uow = new AsyncUnitOfWork();
+            uow.onError((context, phase, error) -> capturedError.set(error)).runOnPreInvocation(context -> {
+                   EventStoreTransaction transaction = defaultEventStoreTransactionFor(context);
+                   transaction.appendEvent(event1);
+               })
+               .runOnPrepareCommit(context -> {
+                   throw new RuntimeException("Simulated failure during prepare commit");
+               }).runOnCommit(context -> onCommitExecuted.set(true))
+               .runOnAfterCommit(context -> onAfterCommitExecuted.set(true))
+               .runOnPostInvocation(context -> onPostInvocationExecuted.set(true));
+
+            RuntimeException exception = assertThrows(CompletionException.class, () -> awaitException(uow.execute()));
+
+            // then
+            assertNotNull(capturedError.get());
+            assertEquals("Simulated failure during prepare commit", capturedError.get().getMessage());
+            assertEquals(exception.getCause(), capturedError.get());
+            assertFalse(onCommitExecuted.get(), "Commit step should not execute after an error");
+            assertFalse(onAfterCommitExecuted.get(), "After commit step should not execute after an error");
+            assertTrue(onPostInvocationExecuted.get(), "Post invocation step should be executed after an error");
+        }
+    }
+
+    // TODO - Discuss: @Steven - Perfect candidate to move to a commons test utils module?
```

Comment:
I thought there was an issue for this, but I can't find it... I left this TODO in the previous test suite, as I would want to ensure we do not copy test-utils, like this, from test class to test class. Instead, it would be beneficial if we gather these, I anticipate per module, in specific place. 

And, importantly, make those part of the deployable.
In doing so, we allow different modules to use the same test utils.

Not something you have to solve right now, nor soon. But, I think it would be nice if we do this somewhere in the future.

Reply:
I will merge that one, but will do it after.